pkg/clusterstatus/etcd/etcd.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2019 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	kubeoneapi "github.com/kubermatic/kubeone/pkg/apis/kubeone"
+	"github.com/kubermatic/kubeone/pkg/ssh"
+	"github.com/kubermatic/kubeone/pkg/state"
+)
+
+const (
+	defaultHTTPTimeout = 10 * time.Second
+	healthEndpoint     = "https://%s:2379/health"
+	membersEndpoint    = "https://127.0.0.1:2379/v2/members"
+)
+
+// Status describes status of the etcd cluster
+type Status struct {
+	Health bool `json:"health,omitempty"`
+	Member bool `json:"member,omitempty"`
+}
+
+type healthRaw struct {
+	Health string `json:"health"`
+}
+
+type membersListRaw struct {
+	Members []struct {
+		ID         string   `json:"id,omitempty"`
+		Name       string   `json:"name,omitempty"`
+		PeerURLs   []string `json:"peerURLs,omitempty"`
+		ClientURLs []string `json:"clientURLs,omitempty"`
+	} `json:"members"`
+}
+
+type HTTPDoer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// EtcdStatus analyzes health of an etcd cluster
+func GetStatus(s *state.State, node kubeoneapi.HostConfig, tunneler ssh.Tunneler) (*Status, error) {
+	client, err := httpClient(s, tunneler)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := membersList(client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check etcd member health
+	healthStr, err := memberHealth(client, node.PrivateAddress)
+	if err != nil {
+		return nil, err
+	}
+	health, err := strconv.ParseBool(healthStr.Health)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check etcd membership
+	var member bool
+	for _, mem := range m.Members {
+		if mem.Name == node.Hostname {
+			member = true
+			break
+		}
+	}
+
+	return &Status{
+		Health: health,
+		Member: member,
+	}, nil
+}
+
+// memberHealth returns health for a requested etcd member
+func memberHealth(c HTTPDoer, nodeAddress string) (*healthRaw, error) {
+	endpoint := fmt.Sprintf(healthEndpoint, nodeAddress)
+	request, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Set("Content-type", "application/json")
+
+	resp, err := c.Do(request)
+	if err != nil {
+		return &healthRaw{Health: "false"}, nil
+	}
```

Comment:
The problem here is if `etcd` is not working, this is going to fail and cause everything to fail instead of not showing `unhealthy`. I should improve error handling, but I'm wondering should I leave it like this for now and fix it in a follow-up to make PR smaller or just go ahead?